tardis/utilities/utils.py

```diff
@@ -66,6 +67,32 @@ def csv_parser(
             }
 
 
+KeyType = TypeVar("KeyType")
+
+
+def deep_update(
+    original_mapping: Mapping[KeyType, Any],
+    mapping_update: Mapping[KeyType, Any],
+) -> Mapping[KeyType, Any]:
+    from collections.abc import Mapping
+
+    updated_mapping = deepcopy(original_mapping)
+    for key, value in mapping_update.items():
+        if isinstance(value, Mapping):
+            updated_mapping[key] = deep_update(value, original_mapping.get(key, {}))
```

Comment:
Not sure if it's sensible to avoid the recursion if we know there is nothing to update.
```suggestion
            updated_mapping[key] = deep_update(original_mapping[key], value) if key in original_mapping else deepcopy(value)
```

Reply:
At any rate, I think the arguments must be inverted. This could affect ordering otherwise.
```suggestion
            updated_mapping[key] = deep_update(original_mapping.get(key, {}), value)
```

Reply:
The first version would preserve types (e.g. `AttributeDict` instead of `dict`) so perhaps that's a good reason to prefer it.